python/mvdtool.cpp

```diff
@@ -108,31 +108,42 @@ namespace {  // Intenal
 template <typename T>
 using pyarray = py::array_t<T, py::array::c_style | py::array::forcecast>;
 
+
+/// Generic routine to copy data item
+template <typename T>
+inline void copy_element(T& dst, const T& src) {
+    dst = src;
+}
+
+/// Copy routine for boost multi_index view to std::array
+template <typename T, size_t Width, typename T2>
+inline void copy_element(std::array<T, Width>& dst, const T2& src) {
+    std::memcpy(&dst[0], src.origin(), sizeof(T) * Width);
+}
+
 /**
  * Extract several elements given their indices from a function accepting a range
  * NOTE: Indexes must be in ascending order
  */
 template <typename T, typename FuncT>
-inline std::vector<T> _atIndices(const FuncT& f,
-                                 const size_t n_records,
-                                 const pyarray<size_t>& idx) {
-    const size_t CHUNK_SIZE = 256;
+inline std::vector<T>_atIndices(const FuncT& f,
+                                const size_t n_records,
+                                const pyarray<size_t>& idx) {
+    constexpr size_t CHUNK_SIZE = 128u;  // 1KB of doubles, more for strings
     const auto indices = idx.template unchecked<1>();
     const auto count = static_cast<size_t>(indices.size());
-    std::vector<T> v(count);
-
-    size_t offset = 0;
-    std::vector<T> chunk;
+    std::vector<T> out_v(count);
 
-    for (size_t i=0; i < count; i++) {
-        const size_t elem_i = indices[i];
-        if(elem_i - offset >= chunk.size()) {
-            offset = elem_i;
-            chunk = f(Range(offset, std::min(CHUNK_SIZE, n_records - offset)));
+    for (size_t i=0; i < count;) {
+        const size_t offset = indices[i];
+        const auto limit = std::min(CHUNK_SIZE, n_records - offset);
+        const auto chunk = f(Range(offset, limit));
```

Comment:
Not exactly sure what type is `chunk` but I guess you can force copy elusion:
```suggestion
        const auto& chunk = f(Range(offset, limit));
```

Reply:
f is supposed to return a constructed value (RVO), but yep, safer and won't hurt!